File: maxSubArraySum.js
// Write a function called maxSubarraySum which accepts an array of integers
// and a number called n. The function should calculate the maximum
// sum of n consecutive elements in the array.

const maxSubarraySum = (arr, num) => {
  let max = 0;
  for (let i = 0; i < num; i++) {
    max += arr[i];
  }
  for (let i = 1; i < arr.length - num; i++) {
    let temp = 0;
    for (let j = i; j < num + i; j++) {
      temp += arr[j];
    }
    max = Math.max(max, temp);
  }
  return max;
};

console.log(maxSubarraySum([1, 2, 5, 2, 8, 1, 5], 4));

// Sliding window approach

const maxSubarraySum2 = (arr, num) => {
  if (arr.length < num) {
    return null;
  }
  let maxSum = 0,
    tempSum = 0;

  for (let i = 0; i < num; i++) {
    maxSum += arr[i];
  }

  tempSum = maxSum;

  for (let i = num; i < arr.length; i++) {
    tempSum += arr[i] - arr[i - num];
    maxSum = Math.max(maxSum, tempSum);
  }
  return maxSum;
};

console.log(maxSubarraySum2([1, 2, 5, 2, 8, 1, 5], 4));

File: python/sliding_window/max_sub_array_sum.py
def max_subarray_sum(arr, window):
    if window > len(arr):
        return None
    maximum = -99999999
    i = 0
    j = 0
    total = 0
    while i <= len(arr) - 1:
        if i == window + j -1:
            i = j
            j += 1
            if total > maximum:
                maximum = total
            total = 0
        total += arr[i]
        i += 1
    return maximum


print(max_subarray_sum([1, 2, 5, 2, 8, 1, 5], 2))

print(max_subarray_sum([1, 2, 5, 2, 8, 1, 5], 4))


# Brute force

def max_subarray_sum_2(arr, window):
    if window > len(arr):
        return None
    maximum = -9999999
    for i in range(len(arr) - window):
        temp = 0
        for j in range(window):
            temp += arr[i + j]
        if temp > maximum:
            maximum = temp
    return maximum

print(max_subarray_sum_2([1, 2, 5, 2, 8, 1, 5], 2))

print(max_subarray_sum_2([1, 2, 5, 2, 8, 1, 5], 4))

            
# Sliding window

def max_subarray_sum_3(arr, window):
    if window > len(arr):
        return None
    max_sum = 0
    temp_sum = 0
    for x in arr[:window]:
        max_sum += x
    temp_sum = max_sum
    for i in range(window, len(arr)):
        temp_sum += (arr[i] - arr[i - window])
        max_sum = max(max_sum, temp_sum)
    return max_sum

print(max_subarray_sum_3([1, 2, 5, 2, 8, 1, 5], 2))

print(max_subarray_sum_3([1, 2, 5, 2, 8, 1, 5], 4))

